### PR TITLE
[IMP] stock: stock.move.line flexible reservation

### PR DESCRIPTION
**Pull request by AhmedAmirK**
This commit allows the user to choose which quants to reserve from stock
and the quantity to reserve from each quant, making reservation much more
flexible from the picking form.
    
In addition, the `reserved_uom_qty` of stock.move.line
is now editable in views to allow for users to change the qty reserved
by existing move lines. When editing the reserved qty, if quantity is not available,
it will reserve only what is available.

taskId: 3090913

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/product_expiry/models/stock_move_line.py [hunk 1]
```diff
@@ -53,3 +53,10 @@ def _prepare_new_lot_vals(self):
         if self.expiration_date:
             vals['expiration_date'] = self.expiration_date
         return vals
+
+    def action_open_reserve_stock(self):
+        res_action = super().action_open_reserve_stock()
+        move_id = self.move_id or self.move_id.browse(self.env.context.get('default_move_id'))
+        if not move_id.product_id.use_expiration_date:
+            res_action['context']['hide_expiry'] = True
```
[the remark this answers is not in the document]
Comment (AhmedAmirK) [reply addons/product_expiry/models/stock_move_line.py:61]: Its not a new context though, its the one passed from the view when creating a move.line. The only scenario the context is used in is when there are no move lines in `self`.

File: addons/stock/wizard/stock_quant_reserve.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            quant_line_cmds += [Command.create({'quant_id': quant.id}) for quant in quant_ids.filtered(lambda q: q.available_quantity > 0)]
+            wiz.quant_line_ids = quant_line_cmds
+
+    def reserve_stock(self):
+        for wiz in self:
+            for line in wiz.quant_line_ids.filtered(lambda l: l.qty_to_reserve > 0):
+                line.quant_id.reserved_quantity += line.qty_to_reserve
+                StockMoveLine = self.env['stock.move.line'].with_context(bypass_reservation_update=True)
+                StockMoveLine.create(wiz.move_id._prepare_move_line_vals(quantity=line.qty_to_reserve, reserved_quant=line.quant_id))
+                wiz.move_id._recompute_state()
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.py:61]: can we move this outside of both of the for loops + batch it with self.move_id._recompute_state()?

File: addons/stock/wizard/stock_quant_reserve.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                                                        ('location_id', 'child_of', move_id.location_id.id)])
+            quant_line_cmds += [Command.create({'quant_id': quant.id}) for quant in quant_ids.filtered(lambda q: q.available_quantity > 0)]
+            wiz.quant_line_ids = quant_line_cmds
+
+    def reserve_stock(self):
+        for wiz in self:
+            for line in wiz.quant_line_ids.filtered(lambda l: l.qty_to_reserve > 0):
+                line.quant_id.reserved_quantity += line.qty_to_reserve
+                StockMoveLine = self.env['stock.move.line'].with_context(bypass_reservation_update=True)
+                StockMoveLine.create(wiz.move_id._prepare_move_line_vals(quantity=line.qty_to_reserve, reserved_quant=line.quant_id))
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.py:60]: can we do the create in batch?

File: addons/stock/wizard/stock_quant_reserve.xml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        <field name="model">stock.quant.reserve</field>
+        <field name="arch" type="xml">
+            <form>
+                <field name="move_id" invisible="1" />
+                <group>
+                    <group>
+                        <label for="qty_to_reserve" />
+                        <div class="o_row">
+                            <span><field name="qty_to_reserve" nolabal="1"/></span>
+                            <span><field name="product_uom_id" nolabal="1"/></span>
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.xml:15]: nolabal? :laughing:

File: addons/stock/wizard/stock_quant_reserve.xml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                        </div>
+                    </group>
+                </group>
+                <div class="oe_grey">
+                    <p>Choose from the following available stock and enter the quantity to reserve.</p>
+                </div>
+                <field name="quant_line_ids" readonly="False" force_save="1">
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.xml:22]: is there a reason why the readonly=False is in the view and not in the field declaration?

File: addons/stock/wizard/stock_quant_reserve.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                    <tree editable="top" create="0">
+                        <field name="quant_id" invisible="1" force_save="1" />
+                        <field name="product_id" />
+                        <field name="location_id" groups="stock.group_stock_multi_locations" />
+                        <field name="lot_id" groups="stock.group_production_lot" invisible="context.get('hide_lot', False)" />
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.xml:27]: don't pass `hide_lot` as a context, add it as a related field in the wizard

File: addons/stock/tests/test_quant_reserve.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        # available should be 0
+        available_qty = self.env['stock.quant']._get_available_quantity(self.product, self.stock_location)
+        self.assertEqual(available_qty, 0)
+        # unreserve, qty available should be 2 and move state back to confirmed
+        move_id.move_line_ids.reserved_uom_qty = 0
+        available_qty = self.env['stock.quant']._get_available_quantity(self.product, self.stock_location)
+        self.assertEqual(available_qty, 2)
+        self.assertEqual(move_id.state, 'confirmed')
+        # reserve qty again, checks to be able to reserve from confirmed state
+        move_id.move_line_ids.reserved_uom_qty = 2
```
[the remark this answers is not in the document]
Comment (AhmedAmirK) [reply addons/stock/tests/test_quant_reserve.py:43]: Good idea

File: addons/stock/tests/test_quant_reserve.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+from odoo.tests.common import TransactionCase
+
+
+class StockQuantReserve(TransactionCase):
```
[the remark this answers is not in the document]
Comment (ticodoo) [reply addons/stock/tests/test_quant_reserve.py:4]: It's still testing quant behavior ;) Also it's strange to create a new file just for 1 test

[the remark this answers is not in the document]
Comment (AhmedAmirK) [reply addons/stock/tests/test_quant_reserve.py:4]: ok

File: addons/stock/wizard/stock_quant_reserve.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
+from odoo import _, api, Command, fields, models
+from odoo.exceptions import UserError
+
+
+class ReserveStockQuantLine(models.TransientModel):
+    _name = 'stock.quant.reserve.line'
+    _description = 'Choose quantity to reserve from each stock quant'
+
+    quant_id = fields.Many2one('stock.quant', required=True, readonly=True)
+    chosen = fields.Boolean()
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.py:10]: This is a slightly cleaner name that is more consistent with our usual boolean field names

```suggestion
    is_selected = fields.Boolean()
```

File: addons/stock/wizard/stock_quant_reserve.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    available_quantity = fields.Float(related='quant_id.available_quantity', string='Available')
+    qty_to_reserve = fields.Float('To reserve', compute='_compute_qty_to_reserve', store=True)
+
+    reserve_id = fields.Many2one('stock.quant.reserve')
+
+    @api.depends('chosen', 'reserve_id', 'available_quantity')
+    def _compute_qty_to_reserve(self):
+        for line in self:
+            if line.chosen and line.reserve_id.qty_to_reserve:
+                line.qty_to_reserve = min(line.reserve_id.qty_to_reserve, line.available_quantity)
```
Comment (ticodoo) [inline addons/stock/wizard/stock_quant_reserve.py:25]: I hope this works

```suggestion
            if line.chosen:
                line.qty_to_reserve = 0
                if line.reserve_id.qty_to_reserve:
                    line.qty_to_reserve = min(line.reserve_id.qty_to_reserve, line.available_quantity)
```

**Comment by amoyaux** [comment]
robodoo r+ rebase-ff

**Comment by amoyaux** [comment]
robodoo r+

**Pull request closed without merging**